File: 2_Programmers/Lv1/[Lv2] 전화번호 목록.py
# 문제: 전화번호 목록 (Lv.2)
# 유형: 해시
# 풀이: 정렬 -> prefix 탐색이 더 효율적

"""
def solution(phone_book):
    s = set(phone_book)

    for phone in phone_book:
        prefix = ""

        for i in phone[:-1]:
            prefix += i
            if prefix in phone_book:
                return False
    return True
"""

def solution(phone_book):
    phone_book.sort()
    for i in range(len(phone_book)-1):
        if phone_book[i+1].startswith(phone_book[i]):
            return False
    return True

File: 2_Programmers/Lv1/[Lv2] 전화번호 목록.java
// 문제: 전화번호 목록 (Lv.2)
// 유형: 해시
// 풀이: 정렬 -> prefix 탐색이 더 효율적

import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        Arrays.sort(phone_book);
        for(int i=0;i<phone_book.length-1;i++){
            if(phone_book[i+1].startsWith(phone_book[i])){
                return false;
            }
        }
        return true;
    }
}